k-distribution/k-tutorial/1_basic/16_collections/README.md

```diff
@@ -1 +1,345 @@
+the `ListItem` symbol), followed by zero or one variables of sort `List`,
+followed by zero or more list elements. An empty list is represented by
+`.List`. These patterns can be matched in `O(log(N))` time. This is the only
+type of `List` pattern that can appear on the left-hand-side of a rule. In
+other words, you are not allowed to write a `List` pattern on the
+left-hand-side with more than one variable of sort `List` in it. You are,
+however, allowed to write such patterns on the right-hand-side of a rule.
+
+Note the meaning of matching on a `List` pattern: a list pattern with no
+variables of sort `List` will match if the list being matched has exactly as
+many elements as `ListItem` symbols in the pattern. It will then match if each
+element in sequence matches the pattern contained in the `ListItem` symbol. A
+list pattern with one variable of sort `List` operates the same way, except
+that it can match any list with at least as many elements as `ListItem`
+symbols, so long as the prefix and suffix of the list match the patterns inside
+the `ListItem` symbols. The variable of sort `List` will be bound to whatever
+elements are left over (`.List` if there are no elements left over).
+
+The `...` syntax is allowed on cells containing lists as well. In this case,
+the meaning of `<cell>... Pattern </cell>` is the same as
+`<cell> _:List (Pattern) </cell>`, the meaning of `<cell> Pattern ...</cell>`
+is the same as `<cell> (Pattern) _:List</cell>`. Because list patterns with
+multiple variables of sort `List` are not allowed, it is an error to write
+`<cell>... Pattern ...</cell>`.
+
+Here is an example of a simple definition that implements a very basic
+function-call semantics using a `List` as a function stack (`lesson-16-a.k`):
+
+```k
+module LESSON-16-B-SYNTAX
+  imports INT-SYNTAX
+  import ID-SYNTAX
+
+  syntax Exp ::= Id "(" ")" | Int
+  syntax Stmt ::= "return" Exp ";" [strict]
+  syntax Decl ::= "fun" Id "(" ")" "{" Stmt "}"
+  syntax Pgm ::= List{Decl,""}
+  syntax Id ::= "main" [token]
+endmodule
+
+module LESSON-16-B
+  imports LESSON-16-B-SYNTAX
+  imports BOOL
+
+  configuration <T>
+                  <k> $PGM:Pgm ~> main () </k>
+                  <functions> .Map </functions>
+                  <fstack> .List </fstack>
+                </T>
+
+  // declaration sequence
+  rule <k> D:Decl P:Pgm => D ~> P ...</k>
+  rule <k> .Pgm => . ...</k>
+
+  // function definitions
+  rule <k> fun X:Id () { S } => . ...</k>
+       <functions>... .Map => X |-> S ...</functions>
+
+  // function call
+  syntax KItem ::= stackFrame(K)
+  rule <k> X:Id () ~> K => S </k>
+       <functions>... X |-> S ...</functions>
+       <fstack> .List => ListItem(stackFrame(K)) ...</fstack>
+
+  // return statement
+  rule <k> return I:Int ; ~> _ => I ~> K </k>
+       <fstack> ListItem(stackFrame(K)) => .List ...</fstack>
+
+  syntax Bool ::= isKResult(K) [function, symbol]
+  rule isKResult(_:Int) => true
+  rule isKResult(_) => false [owise]
```

Comment:
What's the reason we're using this approach rather than `syntax KResult ::= Int`?

Reply:
I would rather introduce people to this syntax first since KResult isn't really intended to be a sort, and we want to discourage people from casting terms to sort KResult, since doing so interacts badly with some features of strictness. It's better than they think of KResult as a function rather than a sort.

Reply:
A total function into Bool is isomorphic to a set, which is what a Sort is in the model theory of matching logic. So I guess I assumed the different definitions were semantically identical. But it sounds like there is some operational difference?

Reply:
I mean, if I say something like:

```
syntax Exp ::= Int | Exp "+" Exp
syntax KResult ::= Int

rule I:KResult + I2:KResult => whatever
```

This will not parse, whereas if I say

```
rule I + I2 => whatever
requires isKResult(I) andBool isKResult(I2)
```

That's fine. Furthermore, if I say:

```
syntax Exp ::= Int | Exp "+" Exp [strict, hybrid]
syntax KResult ::= Int

rule I:KResult => whatever
```

`I` will not match `1 + 2`, even though `isKResult(1 + 2)` will return true.

Thus it's pretty much always better to use isKResult rather than `:KResult`, and I'd prefer to introduce the material in a fashion that encourages that behavior.

Reply:
Hmmm, so we're treating `KResult` specially, in the sense that it's not a proper sort, but a declaration for a function `isKResult`. I guess I don't like that very much, but makes sense.